plugin-server/src/main/ingestion-queues/batch-processing/each-batch-ingestion.ts

```diff
@@ -7,19 +7,50 @@ import { sanitizeEvent } from '../../../utils/utils'
 import { KafkaQueue } from '../kafka-queue'
 import { eachBatch } from './each-batch'
 
-export async function eachMessageIngestion(message: KafkaMessage, queue: KafkaQueue): Promise<void> {
+export function getPluginEvent(message: KafkaMessage): PluginEvent {
+    // TODO: inefficient to do this twice?
     const { data: dataStr, ...rawEvent } = JSON.parse(message.value!.toString())
     const combinedEvent = { ...rawEvent, ...JSON.parse(dataStr) }
     const event: PluginEvent = sanitizeEvent({
         ...combinedEvent,
         site_url: combinedEvent.site_url || null,
         ip: combinedEvent.ip || null,
     })
-    await ingestEvent(queue.pluginsServer, queue.workerMethods, event)
+    return event
+}
+
+export async function eachMessageIngestion(message: KafkaMessage, queue: KafkaQueue): Promise<void> {
+    await ingestEvent(queue.pluginsServer, queue.workerMethods, getPluginEvent(message))
 }
 
 export async function eachBatchIngestion(payload: EachBatchPayload, queue: KafkaQueue): Promise<void> {
-    await eachBatch(payload, queue, eachMessageIngestion, 'ingestion')
+    function groupIntoBatchesIngestion(array: KafkaMessage[], batchSize: number): KafkaMessage[][] {
+        // Once we see a distinct ID we've already seen break up the batch
+        const batches = []
+        const seenIds: Set<string> = new Set()
+        let currentBatch: KafkaMessage[] = []
+        for (const message of array) {
+            const pluginEvent = getPluginEvent(message)
+            const enabledBreakupById = queue.pluginsServer.ingestionBatchBreakupByDistinctIdTeams.has(
+                pluginEvent.team_id
+            )
+            const seenKey = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
+            if (currentBatch.length == batchSize || (enabledBreakupById && seenIds.has(seenKey))) {
+                seenIds.clear()
+                batches.push(currentBatch)
+                currentBatch = []
+            }
+            if (enabledBreakupById) {
+                seenIds.add(seenKey)
+            }
+            currentBatch.push(message)
+        }
+        if (currentBatch) {
+            batches.push(currentBatch)
+        }
+        return batches
+    }
+    await eachBatch(payload, queue, eachMessageIngestion, groupIntoBatchesIngestion, 'ingestion')
```

Comment:
q: are we sure messages in a batch from kafkajs come in order?

Reply:
either way, I believe we should probably explicitly sort by timestamp (and later on also group identifies early).

think about how someone might send a batch to /capture and the order we produce to kafka is not the same as the timestamp ordering. we shouldn't depend on the other service to do this, thus let's sort here

Reply:
hmm fair point